File: scripts/wiki_download.py
#!/usr/bin/env python3
"""
Download a part image from a beyblade.fandom.com wiki URL.

Usage:
    python scripts/wiki_download.py <wiki-url>
    python scripts/wiki_download.py  # prompts for URL

If multiple images are on the page, lists them so you can pick.
"""

import json
import sys
import urllib.parse
import urllib.request
from pathlib import Path

ROOT = Path(__file__).parent.parent
IMG_DIR = ROOT / "public" / "images"
HEADERS = {"User-Agent": "Mozilla/5.0"}
API = "https://beyblade.fandom.com/api.php"


def api_get(params: dict) -> dict:
    qs = urllib.parse.urlencode(params)
    req = urllib.request.Request(f"{API}?{qs}", headers=HEADERS)
    with urllib.request.urlopen(req) as r:
        return json.load(r)


def download(url: str, dest: Path) -> None:
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req) as r, open(dest, "wb") as f:
        f.write(r.read())


def clean_image_url(thumb_url: str) -> str:
    return thumb_url.split("/scale-to-width-down")[0].split("/revision/latest")[0] + "/revision/latest"


def page_title_from_url(url: str) -> str:
    path = urllib.parse.urlparse(url).path
    title = path.split("/wiki/")[-1]
    return urllib.parse.unquote(title).replace("_", " ")


def list_page_images(title: str) -> list[str]:
    data = api_get({"action": "query", "titles": title, "prop": "images", "imlimit": "50", "format": "json"})
    pages = data["query"]["pages"]
    for page in pages.values():
        return [img["title"].removeprefix("File:") for img in page.get("images", [])]
    return []


def get_image_url(filename: str) -> str | None:
    data = api_get({
        "action": "query",
        "titles": f"File:{filename}",
        "prop": "imageinfo",
        "iiprop": "url",
        "format": "json",
    })
    for page in data["query"]["pages"].values():
        info = page.get("imageinfo", [])
        if info:
            return info[0]["url"]
    return None


def main():
    url = sys.argv[1] if len(sys.argv) > 1 else input("Wiki URL: ").strip()
    title = page_title_from_url(url)
    print(f"Page: {title}")

    # Get main page image
    data = api_get({"action": "query", "titles": title, "prop": "pageimages", "pithumbsize": "500", "format": "json"})
    pages = data["query"]["pages"]
    main_image = None
    main_url = None
    for page in pages.values():
        main_image = page.get("pageimage")
        thumb = page.get("thumbnail", {}).get("source", "")
        if thumb:
            main_url = clean_image_url(thumb)

    # List all images on the page
    all_images = list_page_images(title)
    IMAGE_EXTS = {".png", ".jpg", ".jpeg", ".webp", ".gif"}
    blade_images = [i for i in all_images if Path(i).suffix.lower() in IMAGE_EXTS]

    if len(blade_images) > 1:
        print(f"\nFound {len(blade_images)} images on this page:")
        for i, name in enumerate(blade_images):
            marker = " <-- main" if name == main_image else ""
            print(f"  {i}) {name}{marker}")
        choice = input("\nEnter numbers to download (comma-separated, 'all', or Enter = main): ").strip()

        if choice.lower() == "all":
            chosen_images = blade_images
        elif choice:
            indices = [int(x.strip()) for x in choice.split(",") if x.strip().isdigit()]
            chosen_images = [blade_images[i] for i in indices if i < len(blade_images)]
        elif main_image:
            chosen_images = [(main_image, main_url)]  # type: ignore[assignment]
        else:
            print("No image found.")
            return
    elif main_image:
        chosen_images = [main_image]
    else:
        print("No image found on this page.")
        return

    # Resolve (name, url) pairs
    to_download: list[tuple[str, str]] = []
    for item in chosen_images:
        if isinstance(item, tuple):
            to_download.append(item)
        else:
            name = item
            url = main_url if name == main_image and main_url else get_image_url(name)
            if url:
                to_download.append((name, url))
            else:
                print(f"Could not resolve URL for {name}, skipping.")

    for img_name, img_url in to_download:
        dest = IMG_DIR / img_name
        if dest.exists():
            overwrite = input(f"{img_name} already exists. Overwrite? [y/N] ").strip().lower()
            if overwrite != "y":
                print(f"Skipped {img_name}.")
                continue
        print(f"Downloading {img_name} ...", end=" ", flush=True)
        download(img_url, dest)
        size_kb = dest.stat().st_size // 1024
        print(f"OK ({size_kb}KB) → public/images/{img_name}")

    if to_download:
        print("\nAdd to beyparts.js:")
        for img_name, _ in to_download:
            print(f'  image: "{img_name}"')


if __name__ == "__main__":
    main()

File: scripts/wiki_download.js
#!/usr/bin/env node
/**
 * Download a part image from a beyblade.fandom.com wiki URL.
 *
 * Usage:
 *   node scripts/wiki_download.js <wiki-url>
 *   node scripts/wiki_download.js  # prompts for URL
 */

import { createWriteStream, existsSync, statSync } from "fs";
import { createInterface } from "readline";
import { pipeline } from "stream/promises";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const IMG_DIR = path.join(ROOT, "public", "images");
const HEADERS = { "User-Agent": "Mozilla/5.0" };
const API = "https://beyblade.fandom.com/api.php";

const rl = createInterface({ input: process.stdin, output: process.stdout });
const prompt = (q) => new Promise((res) => rl.question(q, res));

async function apiGet(params) {
  const qs = new URLSearchParams(params).toString();
  const res = await fetch(`${API}?${qs}`, { headers: HEADERS });
  return res.json();
}

async function download(url, dest) {
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  await pipeline(res.body, createWriteStream(dest));
}

function cleanImageUrl(thumbUrl) {
  return (
    thumbUrl.split("/scale-to-width-down")[0].split("/revision/latest")[0] +
    "/revision/latest"
  );
}

function pageTitleFromUrl(url) {
  const parsed = new URL(url);
  const title = parsed.pathname.split("/wiki/").pop();
  return decodeURIComponent(title).replace(/_/g, " ");
}

async function listPageImages(title) {
  const data = await apiGet({
    action: "query",
    titles: title,
    prop: "images",
    imlimit: "50",
    format: "json",
  });
  const pages = data.query.pages;
  for (const page of Object.values(pages)) {
    return (page.images ?? []).map((img) =>
      img.title.replace(/^File:/, "")
    );
  }
  return [];
}

async function getImageUrl(filename) {
  const data = await apiGet({
    action: "query",
    titles: `File:${filename}`,
    prop: "imageinfo",
    iiprop: "url",
    format: "json",
  });
  for (const page of Object.values(data.query.pages)) {
    const info = page.imageinfo ?? [];
    if (info.length) return info[0].url;
  }
  return null;
}

async function main() {
  const url =
    process.argv[2] ?? (await prompt("Wiki URL: ")).trim();

  const title = pageTitleFromUrl(url);
  console.log(`Page: ${title}`);

  const data = await apiGet({
    action: "query",
    titles: title,
    prop: "pageimages",
    pithumbsize: "500",
    format: "json",
  });

  const pages = data.query.pages;
  let mainImage = null;
  let mainUrl = null;
  for (const page of Object.values(pages)) {
    mainImage = page.pageimage ?? null;
    const thumb = page.thumbnail?.source ?? "";
    if (thumb) mainUrl = cleanImageUrl(thumb);
  }

  const allImages = await listPageImages(title);
  const IMAGE_EXTS = new Set([".png", ".jpg", ".jpeg", ".webp", ".gif"]);
  const bladeImages = allImages.filter((i) =>
    IMAGE_EXTS.has(path.extname(i).toLowerCase())
  );

  let chosenImages;
  if (bladeImages.length > 1) {
    console.log(`\nFound ${bladeImages.length} images on this page:`);
    bladeImages.forEach((name, i) => {
      const marker = name === mainImage ? " <-- main" : "";
      console.log(`  ${i}) ${name}${marker}`);
    });
    const choice = (
      await prompt(
        "\nEnter numbers to download (comma-separated, 'all', or Enter = main): "
      )
    ).trim();

    if (choice.toLowerCase() === "all") {
      chosenImages = bladeImages;
    } else if (choice) {
      const indices = choice
        .split(",")
        .map((x) => parseInt(x.trim(), 10))
        .filter((n) => !isNaN(n) && n < bladeImages.length);
      chosenImages = indices.map((i) => bladeImages[i]);
    } else if (mainImage) {
      chosenImages = [[mainImage, mainUrl]];
    } else {
      console.log("No image found.");
      rl.close();
      return;
    }
  } else if (mainImage) {
    chosenImages = [mainImage];
  } else {
    console.log("No image found on this page.");
    rl.close();
    return;
  }

  // Resolve (name, url) pairs
  const toDownload = [];
  for (const item of chosenImages) {
    if (Array.isArray(item)) {
      toDownload.push(item);
    } else {
      const name = item;
      const resolvedUrl =
        name === mainImage && mainUrl ? mainUrl : await getImageUrl(name);
      if (resolvedUrl) {
        toDownload.push([name, resolvedUrl]);
      } else {
        console.log(`Could not resolve URL for ${name}, skipping.`);
      }
    }
  }

  for (const [imgName, imgUrl] of toDownload) {
    const dest = path.join(IMG_DIR, imgName);
    if (existsSync(dest)) {
      const overwrite = (
        await prompt(`${imgName} already exists. Overwrite? [y/N] `)
      )
        .trim()
        .toLowerCase();
      if (overwrite !== "y") {
        console.log(`Skipped ${imgName}.`);
        continue;
      }
    }
    process.stdout.write(`Downloading ${imgName} ... `);
    await download(imgUrl, dest);
    const sizeKb = Math.floor(statSync(dest).size / 1024);
    console.log(`OK (${sizeKb}KB) → public/images/${imgName}`);
  }

  if (toDownload.length) {
    console.log("\nAdd to parts-overrides.json:");
    for (const [imgName] of toDownload) {
      console.log(`  "image": "${imgName}"`);
    }
  }

  rl.close();
}

main().catch((e) => {
  console.error(e.message);
  rl.close();
  process.exit(1);
});
